src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs

```diff
@@ -91,23 +94,63 @@ public static bool ShouldHideAdvancedMembers(this Document document)
         }
 
         /// <summary>
-        /// Get the user-specified naming rules, then add standard default naming rules (if provided). The standard 
-        /// naming rules (fallback rules) are added at the end so they will only be used if the user hasn't specified 
-        /// a preference.
+        /// Gets the set of naming rules the user has set for this document.  Will include a set of default naming rules
+        /// that match if the user hasn't specified any for a particular symbol type.  The are added at the end so they
+        /// will only be used if the user hasn't specified a preference.
         /// </summary>
-        internal static async Task<ImmutableArray<NamingRule>> GetNamingRulesAsync(this Document document,
+        public static Task<ImmutableArray<NamingRule>> GetNamingRulesAsync(
+            this Document document, CancellationToken cancellationToken)
+            => document.GetNamingRulesAsync(FallbackNamingRules.Default, cancellationToken);
+
+        /// <summary>
+        /// Get the user-specified naming rules, with the added <paramref name="defaultRules"/>.
+        /// </summary>
+        public static async Task<ImmutableArray<NamingRule>> GetNamingRulesAsync(this Document document,
             ImmutableArray<NamingRule> defaultRules, CancellationToken cancellationToken)
         {
             var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
             var namingStyleOptions = options.GetOption(NamingStyleOptions.NamingPreferences);
             var rules = namingStyleOptions.CreateRules().NamingRules;
 
-            if (defaultRules.Length > 0)
+            return defaultRules.IsDefaultOrEmpty ? rules : rules.AddRange(defaultRules);
+        }
+
+        public static async Task<NamingRule> GetApplicableNamingRuleAsync(this Document document, ISymbol symbol, CancellationToken cancellationToken)
+        {
+            var rules = await document.GetNamingRulesAsync(cancellationToken).ConfigureAwait(false);
+            foreach (var rule in rules)
+            {
+                if (rule.SymbolSpecification.AppliesTo(symbol))
+                    return rule;
+            }
+
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        public static async Task<NamingRule> GetApplicableNamingRuleAsync(
+            this Document document, SymbolKind symbolKind, Accessibility accessibility, CancellationToken cancellationToken)
+        {
+            var rules = await document.GetNamingRulesAsync(cancellationToken).ConfigureAwait(false);
+            foreach (var rule in rules)
+            {
+                if (rule.SymbolSpecification.AppliesTo(symbolKind, accessibility))
+                    return rule;
+            }
+
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        public static async Task<NamingRule> GetApplicableNamingRuleAsync(
```

Comment:
I personally like the style of just having three optional parameters on one method rather than three almost duplicate methods, but I don't know what our general rules are.

Reply:
in this csae, i'm just following the pattern of SymbolSpecification here.  I agree we likely could simplify this greatly by merging together.